Mathlib/Order/Sublattice.lean

```diff
@@ -289,3 +292,97 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
+  prod_mono le_rfl hM
+
+lemma prod_left_mono : Monotone fun L : Sublattice α ↦ L.prod M := fun _ _ ↦ prod_mono_left
+lemma prod_right_mono : Monotone fun M : Sublattice β ↦ L.prod M := fun _ _ ↦ prod_mono_right
+
+lemma prod_top (L : Sublattice α) : L.prod (⊤ : Sublattice β) = L.comap LatticeHom.fst :=
+  ext fun a ↦ by simp [mem_prod, LatticeHom.coe_fst]
+
+lemma top_prod (L : Sublattice β) : (⊤ : Sublattice α).prod L = L.comap LatticeHom.snd :=
+  ext fun a ↦ by simp [mem_prod, LatticeHom.coe_snd]
+
+@[simp] lemma top_prod_top : (⊤ : Sublattice α).prod (⊤ : Sublattice β) = ⊤ :=
+  (top_prod _).trans <| comap_top _
+
+@[simp] lemma prod_bot (L : Sublattice α) : L.prod (⊥ : Sublattice β) = ⊥ :=
+  SetLike.coe_injective prod_empty
+
+@[simp] lemma bot_prod (M : Sublattice β) : (⊥ : Sublattice α).prod M = ⊥ :=
+  SetLike.coe_injective empty_prod
+
+lemma le_prod_iff {M : Sublattice β} {N : Sublattice (α × β)} :
+    N ≤ L.prod M ↔ N ≤ comap LatticeHom.fst L ∧ N ≤ comap LatticeHom.snd M := by
+  simp [SetLike.le_def, forall_and]
+
+@[simp] lemma prod_eq_bot {M : Sublattice β} : L.prod M = ⊥ ↔ L = ⊥ ∨ M = ⊥ := by
+  simpa only [← coe_inj] using Set.prod_eq_empty_iff
+
+@[simp] lemma prod_eq_top [Nonempty α] [Nonempty β] {M : Sublattice β} :
+    L.prod M = ⊤ ↔ L = ⊤ ∧ M = ⊤ := by simpa only [← coe_inj] using Set.prod_eq_univ
+
+/-- The product of sublattices is isomorphic to their product as lattices. -/
+@[simps! toEquiv apply symm_apply]
+def prodEquiv (L : Sublattice α) (M : Sublattice β) : L.prod M ≃o L × M where
+  toEquiv := Equiv.Set.prod _ _
+  map_rel_iff' := Iff.rfl
+
+section Pi
+variable {κ : Type*} {π : κ → Type*} [∀ i, Lattice (π i)]
+
+/-- Arbitrary product of sublattices. Given an index set `s` and a family of sublattices
+`L : Π i, Sublattice (α i)`, `pi s L` is the sublattice of dependent functions `f : Π i, α i` such
+that `f i` belongs to `L i` whenever `i ∈ s`. -/
+@[simps]
+def pi (s : Set κ) (L : ∀ i, Sublattice (π i)) : Sublattice (∀ i, π i) where
+  carrier := s.pi fun i ↦ L i
+  supClosed' := supClosed_pi fun i _ ↦ (L i).supClosed
+  infClosed' := infClosed_pi fun i _ ↦ (L i).infClosed
+
+attribute [norm_cast] coe_pi
+
+@[simp] lemma mem_pi {s : Set κ} {L : ∀ i, Sublattice (π i)} {x : ∀ i, π i} :
+    x ∈ pi s L ↔ ∀ i, i ∈ s → x i ∈ L i := Iff.rfl
+
+@[simp] lemma pi_empty (L : ∀ i, Sublattice (π i)) : pi ∅ L = ⊤ := ext fun a ↦ by simp [mem_pi]
+
+@[simp] lemma pi_top (s : Set κ) : (pi s fun i ↦ ⊤ : Sublattice (∀ i, π i)) = ⊤ :=
+  ext fun a ↦ by simp [mem_pi]
+
+@[simp] lemma pi_bot [Nonempty κ] : (pi univ fun i ↦ ⊥ : Sublattice (∀ i, π i)) = ⊥ :=
+  ext fun a ↦ by simp [mem_pi]
+
+lemma le_pi {s : Set κ} {L : ∀ i, Sublattice (π i)} {M : Sublattice (∀ i, π i)} :
+    M ≤ pi s L ↔ ∀ i ∈ s, M ≤ comap (Pi.evalLatticeHom i) (L i) := by simp [SetLike.le_def]; aesop
+
+@[simp] lemma pi_univ_eq_bot {L : ∀ i, Sublattice (π i)} : pi univ L = ⊥ ↔ ∃ i, L i = ⊥ := by
```

Comment:
Let's be consistent with `univ_pi_eq_empty_iff` here.

Reply:
It's not clear whether it should be `univ_pi_` or `pi_univ_` so I haven't changed that